File: src/main/java/com/timepath/vgui/swing/VGUICanvas.java
package com.timepath.vgui.swing;

import com.timepath.steam.SteamUtils;
import com.timepath.steam.io.VDFNode;
import com.timepath.vgui.Element;
import com.timepath.vgui.ImageUtils;
import com.timepath.vgui.VGUIRenderer;
import java.awt.AlphaComposite;
import java.awt.Color;
import java.awt.Cursor;
import java.awt.Dimension;
import java.awt.Font;
import java.awt.FontMetrics;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.Image;
import java.awt.Point;
import java.awt.Rectangle;
import java.awt.RenderingHints;
import java.awt.event.ComponentAdapter;
import java.awt.event.ComponentEvent;
import java.awt.event.MouseEvent;
import java.awt.event.MouseListener;
import java.awt.event.MouseMotionListener;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import javax.swing.JFrame;
import javax.swing.JPanel;
import javax.swing.SwingUtilities;
import javax.swing.WindowConstants;

/**
 * @author TimePath
 */
@SuppressWarnings("serial")
public class VGUICanvas extends JPanel implements MouseListener, MouseMotionListener {

    private static VDFNode.VDFProperty p(String k, Object v) {
        return new VDFNode.VDFProperty(k, v);
    }

    public static void main(String[] args) throws IOException {
        VGUICanvas canvas = new VGUICanvas();
        VDFNode root = new VDFNode("Root");
        VDFNode node = new VDFNode("Test");
        root.addNode(node);
        node.addAllProperties(new ArrayList<>(Arrays.asList(
                p("enabled", 1),
                p("visible", 1),
                p("xpos", 5),
                p("ypos", 5),
                p("zpos", 0),
                p("wide", 100),
                p("tall", 100),
                p("labeltext", "test"),
                p("textalignment", "center"),
                p("controlname", "label")
        )));
        node = new VDFNode(new FileInputStream(new File(SteamUtils.getSteam(), "resource/FileOpenDialog.res")), StandardCharsets.UTF_8);
        for (VDFNode parent : node.getNodes()) {
            for (VDFNode n : parent.getNodes()) {
                Element e = Element.importVdf(n);
                canvas.r.addElement(e);
            }
        }
        JFrame f = new JFrame();
        f.setContentPane(canvas);
        f.setDefaultCloseOperation(WindowConstants.DISPOSE_ON_CLOSE);
        f.pack();
        f.setLocationRelativeTo(null);
        f.setVisible(true);
    }

    public LinkedList<Element> getElements() {
        return r.getElements();
    }

    public VGUICanvas() {
        addMouseListener(this);
        addMouseMotionListener(this);
        setPreferredSize(new Dimension(640, 480));
        addComponentListener(new ComponentAdapter() {
            @Override
            public void componentResized(ComponentEvent e) {
                resize();
            }

        });
    }

    private void resize() {
        offX = (int) ((getWidth() - r.getInternal().width) / 2);
        offY = (int) ((getHeight() - r.getInternal().height) / 2);
        repaint();
    }

    /**
     * Fired when an element has been dropped
     */
    protected void placed() {
    }

    public void setBackgroundImage(Image background) {
        this.background = background;
        prepareImage(background, this);
        repaint();
    }

    private Rectangle getOutliers() {
        Rectangle rect = new Rectangle(r.getInternal().width, r.getInternal().height);
        for (Element elem : r.getElements()) rect.add(r.bounds(elem));
        return ((Rectangle) (rect));
    }

    /**
     * Convenience method for repainting the bare minimum
     */
    private void doRepaint(Rectangle bounds) {
        r.setElementImage(null);
        repaint(offX + bounds.x, offY + bounds.y, (int) bounds.width - 1, (int) bounds.height - 1);
    }

    private void doRepaint1(Rectangle bounds) {
        r.setElementImage(null);
        repaint(offX + bounds.x, offY + bounds.y, bounds.width, bounds.height);
    }

    private void hover(Element e) {
        if (e.equals(r.getHoveredElement())) return;
// Nothing to do
        // Clean up if needed
        if (r.getHoveredElement() != null)
            doRepaint(r.bounds(r.getHoveredElement()));
        // Draw the new element if needed
        r.setHoveredElement(e);
        if (r != null)
            doRepaint(r.bounds(e));
    }

    /**
     * As soon as the height drops below 480, stops rendering
     */
    private BufferedImage drawGrid() {
        BufferedImage img = new BufferedImage(r.getScreen().width, r.getScreen().height, BufferedImage.TYPE_INT_ARGB);
        Graphics2D g = img.createGraphics();
        g.setComposite(GRID_AC);
        g.setRenderingHint(RenderingHints.KEY_RENDERING, RenderingHints.VALUE_RENDER_SPEED);
        g.setColor(GRID_COLOR);
        int w = r.getScreen().width;
        int h = r.getScreen().height;
        int minGridSpacing = 10;
        int i = minGridSpacing;
        if (i < 0) return ((BufferedImage) (img));
        if (i < 2) {// Optimize for small numbers, stop division by zero
            g.fillRect(0, 0, r.getScreen().width, r.getScreen().height);
            return ((BufferedImage) (img));
        }

        int cross = 0;
        int maxX = w - (w % i);
        int maxY = h - (h % i);
        double multX = r.getScreen().getWidth() / r.getInternal().getWidth();
        double multY = r.getScreen().getHeight() / r.getInternal().getHeight();
        for (int y = -1; y <= (maxY / i); y++) {
            for (int x = -1; x <= (maxX / i); x++) {
                int dx = (int) Math.round((double) (maxX * x * i * multX) / maxX);
                int dy = (int) Math.round((double) (maxY * y * i * multY) / maxY);
                g.drawLine(dx + cross, dy + cross, (int) dx - (1 + cross), (int) dy - (1 + cross));
                g.drawLine((int) dx - (1 + cross), dy + cross, dx + cross, (int) dy - (1 + cross));
            }

        }

        g.dispose();
        return ((BufferedImage) (img));
    }

    @Override
    public void setPreferredSize(Dimension preferredSize) {
        Dimension UISize = new Dimension(preferredSize.width + (2 * offX), preferredSize.height + (2 * offY));
        super.setPreferredSize(UISize);
        currentbg = null;
        gridbg = null;
        r.setScreen(preferredSize);
        //        long gcm = gcm(hudRes.width, hudRes.height)
        double resX = r.getScreen().getWidth();
        double resY = r.getScreen().getHeight();
        double m = resX / resY;
        //        System.out.println(resX + "/" + resY + "=" + m)
        //        System.out.println((resX / gcm) + ":" + (resY / gcm) + " = " + Math.round(m * 480) + "x" + 480)
        r.setInternal(new Dimension((int) Math.round((double) m * 480), 480));
        repaint();
    }

    @Override
    protected void paintComponent(Graphics graphics) {
        Graphics2D g = (Graphics2D) graphics;
        g.setColor(BG_COLOR);
        g.fillRect(0, 0, getWidth(), getHeight());
        if (background != null) {
            if (currentbg != null)
                currentbg = ImageUtils.toCompatibleImage(ImageUtils.resizeImage(background, r.getScreen().width, r.getScreen().height));
            g.drawImage(currentbg, offX, offY, this);
        } else {
            g.setColor(Color.WHITE.darker().darker());
            g.fillRect(offX, offY, (int) Math.round((double) r.getScreen().getWidth() * r.getScale()), (int) Math.round((double) r.getScreen().getHeight() * r.getScale()));
        }

        if (gridbg == null)
            gridbg = ImageUtils.toCompatibleImage(drawGrid());
        g.drawImage(gridbg, offX, offY, this);
        g.drawImage(r.getElementImage(), offX, offY, this);
        if (dragSelecting) {
            g.setComposite(SELECT_AC);
            g.setColor(Color.CYAN.darker());
            g.fillRect(offX + r.getSelectRect().x + 1, offY + r.getSelectRect().y + 1, (int) r.getSelectRect().width - 2, (int) r.getSelectRect().height - 2);
            g.setColor(Color.BLUE);
            g.drawRect(offX + r.getSelectRect().x, offY + r.getSelectRect().y, (int) r.getSelectRect().width - 1, (int) r.getSelectRect().height - 1);
        }

    }

    @Override
    public void mouseDragged(MouseEvent e) {
        Point p = e.getPoint();
        p.translate(-offX, -offY);// Localize
        if (SwingUtilities.isLeftMouseButton(e)) {
            if (dragSelecting) {
                r.select(dragStart, p, e.isControlDown());
            } else if (dragMoving) {
                setCursor(Cursor.getPredefinedCursor(Cursor.MOVE_CURSOR));
                r.setDragX((int) r.getDragX() + p.x - dragStart.x);
                r.setDragY((int) r.getDragY() + p.y - dragStart.y);
                r.setElementImage(null);
                repaint();
                dragStart = p;
            }

        }

    }

    @Override
    public void mouseMoved(MouseEvent e) {
        Point p = e.getPoint();
        p.translate(-offX, -offY);// Localize
        hover(chooseBest(r.pick(p)));
    }

    @Override
    public void mousePressed(MouseEvent e) {
        Point p = e.getPoint();
        p.translate(-offX, -offY);// Localize
        if (SwingUtilities.isLeftMouseButton(e)) {
            dragStart = new Point(p.x, p.y);
            r.getSelectRect().setSize(p.x, p.y);
            if (r.getHoveredElement() == null) {// Clicked nothing
                if (!e.isControlDown()) r.deselectAll();
                dragSelecting = true;
                dragMoving = false;
            } else {// Hovering over something
                dragSelecting = false;
                dragMoving = true;
                if (e.isControlDown()) {
                    // Toggle select
                    if (r.isSelected(r.getHoveredElement())) {
                        r.deselect(r.getHoveredElement());
                    } else {
                        r.select(r.getHoveredElement());
                    }

                } else {
                    // If the thing I'm hovering isn't selected already, drop all selections
                    if (!r.isSelected(r.getHoveredElement())) {
                        r.deselectAll();
                        r.select(r.getHoveredElement());
                    }

                }

            }

        }

    }

    @Override
    public void mouseReleased(MouseEvent e) {
        if (SwingUtilities.isLeftMouseButton(e)) {
            setCursor(Cursor.getDefaultCursor());
            if (dragMoving) placed();// Release element
            dragSelecting = false;
            dragMoving = false;
            dragStart = null;
            Rectangle original = new Rectangle(r.getSelectRect());
            r.getSelectRect().setSize(0, 0);
            for (Element elem : r.getSelectedElements()) {
                // ???
                if (r.getSelectedElements().contains(elem.getParent()) && !elem.getParent().getName().replaceAll("\"", "").endsWith(".res")) {// XXX: hacky
                    continue;
                }

                r.translate(elem, r.getDragX(), r.getDragY());
            }

            r.setDragX(0);
            r.setDragY(0);
            doRepaint1(original);
        }

    }

    @Override
    public void mouseClicked(MouseEvent e) {
    }

    @Override
    public void mouseEntered(MouseEvent e) {
    }

    @Override
    public void mouseExited(MouseEvent e) {
    }

    private static Element chooseBest(List<Element> potential) {
        Element smallest = null;
        for (Element elem : potential) {
            if (smallest == null) {
                smallest = elem;
            } else if (elem.getLayer() > smallest.getLayer()) {// Sort by layer, then by size
                smallest = elem;
            } else if (elem.getLayer() == smallest.getLayer() && elem.getSize() < smallest.getSize()) {
                smallest = elem;
            }

        }

        return smallest;
    }

    public VGUIRenderer getR() {
        return r;
    }

    public void setR(VGUIRenderer r) {
        this.r = r;
    }

    private static final AlphaComposite GRID_AC = AlphaComposite.getInstance(AlphaComposite.SRC_OVER, 0.25f);
    private static final AlphaComposite SELECT_AC = AlphaComposite.getInstance(AlphaComposite.SRC_OVER, 0.25f);
    private Image background;
    private BufferedImage currentbg;
    private BufferedImage gridbg;
    private Color BG_COLOR = Color.GRAY;
    private Color GRID_COLOR = Color.WHITE;
    /**
     * Left
     */
    private int offX = 10;
    /**
     * Top
     */
    private int offY = 10;
    /**
     * Initial drag point
     */
    private Point dragStart;
    private boolean dragSelecting;
    private boolean dragMoving;
    private VGUIRenderer r = new VGUIRenderer() {
        @Override
        public void doRepaint(final Rectangle bounds) {
            VGUICanvas.this.doRepaint(bounds);
        }

        @Override
        public void repaint() {
            VGUICanvas.this.repaint();
        }

        @Override
        public FontMetrics getFontMetrics(final Font font) {
            return VGUICanvas.this.getFontMetrics(font);
        }

    };
}

File: src/main/kotlin/com/timepath/vgui/swing/VGUICanvas.kt
package com.timepath.vgui.swing

import com.timepath.steam.SteamUtils
import com.timepath.steam.io.VDFNode
import com.timepath.steam.io.VDFNode.VDFProperty
import com.timepath.vgui.Element
import com.timepath.vgui.ImageUtils
import com.timepath.vgui.VGUIRenderer
import java.awt.*
import java.awt.event.*
import java.awt.image.BufferedImage
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.util.ArrayList
import java.util.Arrays
import java.util.LinkedList
import javax.swing.JFrame
import javax.swing.JPanel
import javax.swing.SwingUtilities
import javax.swing.WindowConstants
import kotlin.platform.platformStatic

/**
 * @author TimePath
 */
SuppressWarnings("serial")
public class VGUICanvas : JPanel(), MouseListener, MouseMotionListener {

    public val r: VGUIRenderer = object : VGUIRenderer() {
        override fun doRepaint(bounds: Rectangle) = this@VGUICanvas.doRepaint(bounds)
        override fun repaint() = this@VGUICanvas.repaint()
        override fun getFontMetrics(font: Font) = this@VGUICanvas.getFontMetrics(font)
    }

    public val elements: LinkedList<Element> get() = r.elements

    override fun mouseClicked(e: MouseEvent) = Unit
    override fun mouseEntered(e: MouseEvent) = Unit
    override fun mouseExited(e: MouseEvent) = Unit

    /** Fired when an element has been dropped */
    protected fun placed(): Unit = Unit

    init {
        addMouseListener(this)
        addMouseMotionListener(this)
        setPreferredSize(Dimension(640, 480))
        addComponentListener(object : ComponentAdapter() {
            override fun componentResized(e: ComponentEvent) = resize()
        })
    }

    private fun resize() {
        offX = ((getWidth() - r.internal.width) / 2).toInt()
        offY = ((getHeight() - r.internal.height) / 2).toInt()
        repaint()
    }

    public fun setBackgroundImage(background: Image) {
        this.background = background
        prepareImage(background, this)
        repaint()
    }

    private fun getOutliers(): Rectangle {
        val rect = Rectangle(r.internal.width, r.internal.height)
        r.elements.forEach { rect.add(r.bounds(it)) }
        return rect
    }

    /** Convenience method for repainting the bare minimum */
    internal fun doRepaint(bounds: Rectangle, add: Int = 0) {
        r.elementImage = null
        repaint(offX + bounds.x, offY + bounds.y, bounds.width - 1 + add, bounds.height - 1 + add)
    }

    private fun hover(elem: Element?) {
        val current = r.hoveredElement
        if (elem == current) return
        r.hoveredElement = elem
        current?.let { doRepaint(r.bounds(it)) }
        elem?.let { doRepaint(r.bounds(it)) }
    }

    /** As soon as the height drops below 480, stops rendering */
    private fun drawGrid(): BufferedImage {
        val img = BufferedImage(r.screen.width, r.screen.height, BufferedImage.TYPE_INT_ARGB)
        val g = img.createGraphics()
        g.setComposite(GRID_AC)
        g.setRenderingHint(RenderingHints.KEY_RENDERING, RenderingHints.VALUE_RENDER_SPEED)
        g.setColor(GRID_COLOR)
        val w = r.screen.width
        val h = r.screen.height
        val minGridSpacing = 10
        val i = minGridSpacing
        if (i < 0) return img
        if (i < 2) {
            // Optimize for small numbers, stop division by zero
            g.fillRect(0, 0, r.screen.width, r.screen.height)
            return img
        }

        val cross = 0
        val maxX = w - (w % i)
        val maxY = h - (h % i)
        val multX = r.screen.getWidth() / r.internal.getWidth()
        val multY = r.screen.getHeight() / r.internal.getHeight()
        for (y in -1..(maxY / i)) {
            for (x in -1..(maxX / i)) {
                val dx = Math.round((maxX.toDouble() * x.toDouble() * i.toDouble() * multX).toDouble() / maxX.toDouble()).toInt()
                val dy = Math.round((maxY.toDouble() * y.toDouble() * i.toDouble() * multY).toDouble() / maxY.toDouble()).toInt()
                g.drawLine(dx + cross, dy + cross, dx.toInt() - (1 + cross), dy.toInt() - (1 + cross))
                g.drawLine(dx.toInt() - (1 + cross), dy + cross, dx + cross, dy.toInt() - (1 + cross))
            }
        }
        g.dispose()
        return img
    }

    override fun setPreferredSize(preferredSize: Dimension) {
        val UISize = Dimension(preferredSize.width + (2 * offX), preferredSize.height + (2 * offY))
        super<JPanel>.setPreferredSize(UISize)
        currentbg = null
        gridbg = null
        r.screen = preferredSize
        //        long gcm = gcm(hudRes.width, hudRes.height)
        val resX = r.screen.getWidth()
        val resY = r.screen.getHeight()
        val m = resX / resY
        //        System.out.println(resX + "/" + resY + "=" + m)
        //        System.out.println((resX / gcm) + ":" + (resY / gcm) + " = " + Math.round(m * 480) + "x" + 480)
        r.internal = Dimension(Math.round(m.toDouble() * 480).toInt(), 480)
        repaint()
    }

    override fun paintComponent(g: Graphics) = (g as Graphics2D).let { g ->
        g.setColor(BG_COLOR)
        g.fillRect(0, 0, getWidth(), getHeight())
        if (background != null) {
            if (currentbg != null)
                currentbg = ImageUtils.toCompatibleImage(ImageUtils.resizeImage(background!!, r.screen.width, r.screen.height))
            g.drawImage(currentbg, offX, offY, this)
        } else {
            g.setColor(Color.WHITE.darker().darker())
            g.fillRect(offX, offY, Math.round(r.screen.getWidth().toDouble() * r.scale).toInt(), Math.round(r.screen.getHeight().toDouble() * r.scale).toInt())
        }

        if (gridbg == null)
            gridbg = ImageUtils.toCompatibleImage(drawGrid())
        g.drawImage(gridbg, offX, offY, this)
        g.drawImage(r.elementImage, offX, offY, this)
        if (dragSelecting) {
            g.setComposite(SELECT_AC)
            g.setColor(Color.CYAN.darker())
            g.fillRect(offX + r.selectRect.x + 1, offY + r.selectRect.y + 1, r.selectRect.width.toInt() - 2, r.selectRect.height.toInt() - 2)
            g.setColor(Color.BLUE)
            g.drawRect(offX + r.selectRect.x, offY + r.selectRect.y, r.selectRect.width.toInt() - 1, r.selectRect.height.toInt() - 1)
        }
    }

    override fun mouseDragged(e: MouseEvent) {
        val p = e.getPoint().let {
            it.translate(-offX, -offY) // Localize
            it
        }
        if (SwingUtilities.isLeftMouseButton(e)) {
            if (dragSelecting) {
                r.select(dragStart!!, p, e.isControlDown())
            } else if (dragMoving) {
                setCursor(Cursor.getPredefinedCursor(Cursor.MOVE_CURSOR))
                r.dragX = r.dragX.toInt() + p.x - dragStart!!.x
                r.dragY = r.dragY.toInt() + p.y - dragStart!!.y
                r.elementImage = null
                repaint()
                dragStart = p
            }
        }
    }

    override fun mouseMoved(e: MouseEvent) {
        val elem = r.pick(e.getPoint().let {
            it.translate(-offX, -offY) // Localize
            it
        }).let { chooseBest(it) }
        hover(elem)
    }

    override fun mousePressed(e: MouseEvent) {
        val p = e.getPoint().let {
            it.translate(-offX, -offY) // Localize
            it
        }
        if (SwingUtilities.isLeftMouseButton(e)) {
            dragStart = Point(p.x, p.y)
            r.selectRect.setSize(p.x, p.y)
            if (r.hoveredElement == null) {
                // Clicked nothing
                if (!e.isControlDown()) r.deselectAll()
                dragSelecting = true
                dragMoving = false
            } else {
                // Hovering over something
                dragSelecting = false
                dragMoving = true
                if (e.isControlDown()) {
                    // Toggle select
                    if (r.isSelected(r.hoveredElement!!)) {
                        r.deselect(r.hoveredElement)
                    } else {
                        r.select(r.hoveredElement)
                    }

                } else {
                    // If the thing I'm hovering isn't selected already, drop all selections
                    if (!r.isSelected(r.hoveredElement!!)) {
                        r.deselectAll()
                        r.select(r.hoveredElement)
                    }
                }
            }
        }
    }

    override fun mouseReleased(e: MouseEvent) {
        if (SwingUtilities.isLeftMouseButton(e)) {
            setCursor(Cursor.getDefaultCursor())
            if (dragMoving) placed() // Release element
            dragSelecting = false
            dragMoving = false
            dragStart = null
            val original = Rectangle(r.selectRect)
            r.selectRect.setSize(0, 0)
            for (elem in r.getSelectedElements()) {
                // ???
                if (elem.parent in r.getSelectedElements() && !elem.parent!!.name!!.replaceAll("\"", "").endsWith(".res")) {
                    // XXX: hacky
                    continue
                }
                r.translate(elem, r.dragX.toDouble(), r.dragY.toDouble())
            }
            r.dragX = 0
            r.dragY = 0
            doRepaint(original, 1)
        }
    }


    private var background: Image? = null
    private var currentbg: BufferedImage? = null
    private var gridbg: BufferedImage? = null
    private val BG_COLOR = Color.GRAY
    private val GRID_COLOR = Color.WHITE
    /** Left */
    private var offX = 10
    /** Top */
    private var offY = 10
    /** Initial drag point */
    private var dragStart: Point? = null
    private var dragSelecting: Boolean = false
    private var dragMoving: Boolean = false

    companion object {

        public platformStatic fun main(args: Array<String>) {
            val canvas = VGUICanvas()
            val root = VDFNode("Root")
            VDFNode("Test").let {
                fun p(k: String, v: Any) = VDFNode.VDFProperty(k, v)
                it.addAllProperties(listOf(
                        p("enabled", 1),
                        p("visible", 1),
                        p("xpos", 5),
                        p("ypos", 5),
                        p("zpos", 0),
                        p("wide", 100),
                        p("tall", 100),
                        p("labeltext", "test"),
                        p("textalignment", "center"),
                        p("controlname", "label")
                ))
                it
            }.let {
                root.addNode(it)
            }
            FileInputStream(File(SteamUtils.getSteam(), "resource/FileOpenDialog.res")).use {
                VDFNode(it, Charsets.UTF_8).let {
                    it.getNodes().forEach {
                        it.getNodes().forEach {
                            Element.importVdf(it).let {
                                canvas.r.addElement(it)
                            }
                        }
                    }
                }
            }
            val f = JFrame().let {
                it.setDefaultCloseOperation(WindowConstants.DISPOSE_ON_CLOSE)
                it.setContentPane(canvas)
                it.pack()
                it.setLocationRelativeTo(null)
                it
            }
            f.setVisible(true)
        }

        private fun chooseBest(potential: List<Element>): Element? {
            if (potential.isEmpty()) return null
            return potential.fold(potential.first(), { smallest: Element, it: Element ->
                when {
                    it.layer > smallest.layer -> it
                    it.layer == smallest.layer && it.size < smallest.size -> it
                    else -> smallest
                }
            })
        }

        private val GRID_AC = AlphaComposite.getInstance(AlphaComposite.SRC_OVER, 0.25f)
        private val SELECT_AC = AlphaComposite.getInstance(AlphaComposite.SRC_OVER, 0.25f)
    }
}
